File: 1286-constrained-subsequence-sum/constrained-subsequence-sum.java
class Solution {
    public int constrainedSubsetSum(int[] nums, int k){
        int n = nums.length;
        int[] dp = new int[n];
        for(int i = 0; i < n; i++){
            dp[i] = nums[i];
        }
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> b[0] - a[0]);
        pq.offer(new int[]{nums[0], 0});
        int ans = dp[0];
        for(int i = 1; i < n; i++){
            while(!pq.isEmpty() && i - pq.peek()[1] > k){
                pq.poll();
            }
            dp[i] = Math.max(dp[i], nums[i] + pq.peek()[0]);
            pq.offer(new int[]{dp[i], i});
            ans = Math.max(ans, dp[i]);
        }
        return ans;
    }
}

File: 1286-constrained-subsequence-sum/constrained-subsequence-sum.py
class Solution(object):
    def constrainedSubsetSum(self, nums, k):
        n = len(nums)
        dp = nums[:]
        pq = []
        heapq.heappush(pq, (-nums[0], 0))
        ans = dp[0]
        for i in range(1, n):
            while pq and i - pq[0][1] > k:
                heapq.heappop(pq)
            dp[i] = max(dp[i], nums[i] - pq[0][0])
            heapq.heappush(pq, (-dp[i], i))
            ans = max(ans, dp[i])
        return ans

        